### ISSUE TITLE
Differentiate between cue and cue-sized notes

### ISSUE THREAD

**Issue by craigsapp** [music-encoding/music-encoding#926]
Issue https://github.com/rism-digital/verovio/issues/2625 highlights that there are two semantically different types of notes that currently are lumped together with `note@cue="true"`.   There should be an encoding method to differentiate between these two types of visually similar note types:

* cue notes, which are small notes that should not be played by a performer.
* cue-sized notes which should be played by the performer.

One practical difference is that cue notes should have cue-size lyrics, while cue-size notes should have regular-sized lyrics.

I propose adding the attribute `<note@size>` to handle the second case.   The main units would be in percentages of the default size (probably dependent on if the note is regular, cue or grace).   For example `<note@size="80%">` would make the note 80% of its normal size.

There should be two additional symbolic sizes allowed in `<note@size>`: `cue` and `grace`.  These would in turn be controllable by some `staffDef` or `scoreDef` attribute/element that allows specifying the  size scaling for true cue and grace notes that could be shared with cue-like and grace-like notes that instead use `note@size` to control their size.

In the future, there will probably be a need for a separate size control `<note@head.size>` .  This controls the notehead size only, without adjusting the size (length) of the stem.  So consideration for how `<note@size>` and `<note@head.size>` interact should be made sooner rather than later.

--------------

And alternate possibility is to mirror how `note@grace` works: In this case there are three categories: `acc`, `unacc` and `unknown`.  For `note@cue`, there could be `cue` and `cue-like`.

**Comment by lpugin** [comment]
It looks good to me to have `note@size` with percent. However, I would avoid allowing additional symbolic values. Beside the fact that it makes typed implementation such as in Verovio complicated, the main problem I see is the confusion it would yield with `@grace` (and in particular `@grace="unkown"`).

Instead of the proposed `note@cue="cue-like"`, which maintains the undesirable overlap between cues and small notes,  what about a distinct boolean attribute such as `@fioritura`? That way we distinguish the semantic role of the notes and their size. I can see three main categories for the role:
* `@grace` for the notes having their duration taken from the surrounding notes
* `@fioritura` (or another name) for notes having their own durations
* `@cue` for notes of another part

**Comment by rettinghaus** [comment]
I would propose to drop `@cue` for layer elements entirely and only have it on layer. 
The problems we face right now are mixed encodings of "normal" notes and cue notes that are totally unclear, e.g., 

```xml
<beam>
  <note />
  <note @cue="true" />
</beam>
```

Would that imply the beam is also cue-sized? If we put `@cue="true"` on the beam, does it mean all notes in there are actually cue notes? 
As "real" cue notes are coming from another instrument it makes sense to have it available on `<layer>`. If there is a cued layer only partially visible, one can use `<space>` like in un-cued layers. 

Furthermore, having `@cue` only on *some* layer elements, and not available on any control element is another problem. 
Cue layers often contain dynamics, slurs / ties and fermatas. How would you encode these to be actually cue-sized? 
If the whole layer shows a cue you can assign the control element in question directly to the cued layer with the `@layer` attribute and imply the size. 

Specifying the actual (relative) size for every single visual thing of a score still is something needed here.

**Comment by craigsapp** [comment]
I like `@fioritura`, but not the name.  For one thing, fioritura is a more restricted concept than "cue-like":

> an embellishment of a melody, especially as improvised by an operatic singer.

Also, fioritura does not necessarily have to be written in cue-sized notes, for example, and as it is traditionally improvised, it may not even be written down.

Here is original (middle staff) and ornamented version (top staff) of the opening bars of Corelli's Violin Sonata Op. 5, No. 1, shown in the appendix of Estienne Roger's 1710 print (Amsterdam):

<img width="1155" alt="Screen Shot 2022-03-14 at 2 51 28 AM" src="https://user-images.githubusercontent.com/3487289/158147982-5744b056-d75d-4aec-93ec-b16eca1284d0.png">

(each performer improvises their own ornamentation of the written music).



In issue https://github.com/rism-digital/verovio/issues/2625#issuecomment-1065884122 there is this case:

<img width="425" alt="Screen Shot 2022-03-14 at 2 43 13 AM" src="https://user-images.githubusercontent.com/3487289/158146575-9deee934-3e88-461f-9bca-602002dcb2f8.png">

Which is a cue-sized note that is not a melodic embellishment.

**Comment by craigsapp** [comment]
> I would propose to drop `@cue` for layer elements entirely and only have it on layer.

That seems like an interesting idea, although a survey of how part cues are handled in a wide range of orchestral and piano parts would be good to do to see if this would limit certain existing cases.   

For cue-like (cue-sized) notes, a layer-level setting would be nice for Chopin 28/8:

<img width="858" alt="Screen Shot 2022-03-14 at 2 13 07 AM" src="https://user-images.githubusercontent.com/3487289/158141593-a42fd37d-443f-44aa-9f3f-1be33e70126e.png">

**Comment by craigsapp** [comment]
> Would that imply the beam is also cue-sized? 

In that case the beam must be regular sized since there is a real note that the instrument must play.

> Would that imply the beam is also cue-sized? If we put `@cue="true"` on the beam, does it mean all notes in there are actually cue notes?

I would say so: If there is an ancestor element containing `@cue="true"` then the notes should be considered cue.   This is an extension of `layer@cue="true"`.   If any descendent notes should not be cue for some reason, then they should have `note@cue="false"`.

**Comment by lpugin** [comment]
> I like `@fioritura`, but not the name.

I agree it is not really appropriate. Maybe simply `@small` ? That might seem dull, but maybe a more neutral solution for small notes that are not a grace note and not a cue (i.e., content of another part).

**Comment by craigsapp** [comment]
`@small` is better than `@fioritura` although it seems too neutral since these notes should be the same size as cue notes.  But if the values of `@small` are `true` and `false`, there should not be any significant problem since the renderer can infer that small notes are the same size as cue notes.  They should not have the size of grace notes for example, which are usually smaller than cue notes (see https://github.com/rism-digital/verovio/issues/586)

**Comment by ahankinson** [comment]
And what about combinations of `@grace` and `@small`? Or `@cue`?  Are they even smaller still?

Putting `@small` in with no other explanation of what it’s function is, is not really great. Should we have a corresponding `@large` too?

The risk, of course, is that we start marking things up by their visual appearance, rather than styling them appropriately based on their function. We are mixing content and presentation here.

How about segmenting the size to the `@glyph.` set of attributes? Adding `@glyph.size`?

At least then we know that it’s specifically about the symbol being used to draw it, rather than any inherent quality of the note itself.

**Comment by lpugin** [comment]
> And what about combinations of `@grace` and `@small?` Or `@cue`? Are they even smaller still?

That is something we can prevent with Schematron. At least to prevent `@small` and `@grace` to be used together. I am not sure what is typically happening with grace notes in cues. Maybe these are indeed even smaller.

I don't find `@small` a very good attribute name, but `@glyph.size` is not any better. The rendering of a note is not necessarily or exclusively rendered with a glyph - which is another reason why I find `@fontsize` not good either.

**Comment by craigsapp** [comment]
> Maybe these are indeed even smaller.

A note can be both a cue note (or cue-like note) at the same time as being a grace note, and the cue-grace note is smaller than a regular grace note.

**Comment by craigsapp** [comment]
> And what about combinations of `@grace` and `@small`? Or `@cue`? Are they even smaller still?

Yes, roughly a regular grace note is slightly smaller than a cue note.   Roughly 70% for a cue-sized note and 65% for a grace note.   Cues can contain grace notes, in which case they are about 50% the size of regular notes.  Example from the above-mentioned issue:

<img width="878" alt="Screen Shot 2022-03-14 at 1 54 56 PM" src="https://user-images.githubusercontent.com/3487289/158259727-ccceaf0e-e9b6-4150-934e-814a58716bbc.png">

Where grace notes are displayed at 65% regular size when used with regular notes, and displayed at 50% regular size when displayed with 70% sized cue notes.   

Note that grace notes can be inside of cue music, but not the other way around.

> Putting `@small` in with no other explanation of what it’s function is, is not really great. Should we have a corresponding `@large` too?

There is no common notational concept of a large note, so `@small="false"` would imply  a regular sized note.

> The risk, of course, is that we start marking things up by their visual appearance, rather than styling them appropriately based on their function. We are mixing content and presentation here.

That is the general reservation about the term `small`.   Cue and cue-sized notes will be the same size, but that size is flexible (say from about 65% to 75% of a regular note size), depending on the house style.

> How about segmenting the size to the `@glyph`. set of attributes? Adding `@glyph.size`?

No, "cue", "cue-like" and "grace" are semantic properties since they have functions associated more closely than just the size of the note:

* grace note: has no score duration and steals time from an adjacent note (either before or after, or may even affect the tempo to squeeze them in).
* cue note: a cue note is a note that has duration but is given as a cue to a performer to help them not get lost during a performance (it shows them where a different part is playing when they are resting).
* cue-sized notes: are visually identical to cue notes, but they are played by the performer (i.e., they have a very different semantic meaning even though they look the same).

* lyrics for cue notes are cue-sized and lyrics for cue-sized notes are regular size
* grace notes can occur within cue and cue-sized notes (and they are smaller than these notes in that case so that they can be identified as grace notes).   These are not very common, however. 
* Typically cue notes will be found in one layer, while a regular size rest is found in another layer (this is the performer's part).  Cue-sized notes are found by themselves without a paired regular-sized rest.     Here is an exception that I came across:

https://kern.humdrum.org/cgi-bin/ksdata?file=erk047.krn&l=users/craig/songs/erk/liederschatz/band1&format=pdf

<img width="1278" alt="Screen Shot 2022-03-11 at 3 05 32 AM" src="https://user-images.githubusercontent.com/3487289/157855468-9cc5757f-a87a-423f-b6f4-20bb92ad8707.png">

The notes are "cue-sized", but not "cue" since it is expected for the performer to sing them.  But contrary to most cue-sized notation, there is a large rest which makes them otherwise look like cue notes. (The `p` dynamic is also cue-sized).   

The "la la la" lyrics do not seem to be smaller than the other lyrics, which is also a property more associated with "cue-size" rather than "cue" notes.

**Comment by lpugin** [comment]
I would still try to avoid `cue-sized` if possible. What about `ornam-sized`?

Obviously, any choice need to come with some explanation for what it is meant for. That should include the fact that `@cue-sized` (or `@ornam-sized`, or whatever) on notes would not have any impact on lyric size.

**Comment by craigsapp** [comment]
Here is some discussion about cue and small notes, cadenzas: https://github.com/music-encoding/music-encoding/issues/406

**Comment by craigsapp** [comment]
> What about `ornam-sized`?

I don't think so.  What would that mean?

Behind Bars page 471 mentions "cue-sized":

<img width="793" alt="Screen Shot 2022-03-14 at 11 43 21 PM" src="https://user-images.githubusercontent.com/3487289/158325916-c9e1e63f-7ed8-4195-973d-51c8b3f3e6d6.png">

but otherwise, I have not found any other mention about this sort of small non-cue note in Behind Bars.

**Comment by craigsapp** [comment]
Page 568  in Behind Bars discussing lyrics cues:

<img width="720" alt="Screen Shot 2022-03-15 at 12 20 12 AM" src="https://user-images.githubusercontent.com/3487289/158326564-7f64932b-eaee-4105-bc31-aba8dbaa3b10.png">

**Comment by lpugin** [comment]
And an example of a grace note within a cue

![image](https://user-images.githubusercontent.com/689412/159228690-3f0b3c1e-b51e-4c12-9c0f-91491b9ef6bf.png)

**Comment by bwbohl** [comment]
Pushing this to MEI 5.x milestone because the associated PR is still a draft and discussion points to additional thought that has to be put into this matter.
